src/com/grundszok/piotr/app/controllers/MenuController.java

```diff
@@ -58,42 +62,51 @@ public void start() {
             final boolean listValid = shoppingList == null || shoppingList.isEmpty();
             switch (inputService.getUserChoice()) {
                 case "1" -> {
-                    print("Podaj pierwszy produkt:");
+                    displayService.print("Podaj pierwszy produkt:");
                     this.shoppingList = inputService.getShopItemsFromUser(this.shoppingList);
                 }
                 case "2" -> {
                     if (listValid) {
-                        print(INVALID_LIST);
+                        displayService.print(INVALID_LIST);
                         break;
                     }
 
                     String productCollection = this.shoppingList.stream().map(Object::toString).collect(Collectors.joining("\n"));
-                    print(productCollection);
+                    displayService.print(productCollection);
                     displayService.printWithoutStyle("\n\nWprowadz dowolny przycisk zeby przejsc do menu");
                     inputService.getUserChoice();
                 }
                 case "3" -> {
                     if (listValid) {
-                        print(INVALID_LIST);
+                        displayService.print(INVALID_LIST);
                         break;
                     }
-                    print("Czy jestes pewien ze chcesz wyczyscic liste zakupow? Jesli tak, wpisz \"3\"");
+                    displayService.print("Czy jestes pewien ze chcesz wyczyscic liste zakupow? Jesli tak, wpisz \"3\"");
                     if (inputService.getUserChoice().equals("3")) {
                         this.shoppingList.clear();
-                        print("Wyczyszczono liste zakupów");
+                        displayService.print("Wyczyszczono liste zakupów");
                     }
                 }
                 case "4" -> chooseDisplayStrategyFromUserInput();
+                case "5" -> {
+                    try {
+                        persistanceService.save(shoppingList);
+                        displayService.print(String.format("Zapisano %d elementów do pliku", shoppingList.size()));
+                    } catch (EmptyShoppingListException exception) {
+                        displayService.print(exception.getMessage());
+                    }
+                }
                 case "X" -> System.exit(0);
```

Comment:
Move strings to proper class and refactor cases to use private methods for readability sake